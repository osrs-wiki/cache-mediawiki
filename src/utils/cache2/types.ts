export enum IndexType {
  Animations = 0,
  Skeletons = 1,
  Configs = 2,
  Interfaces = 3,
  SoundEffects = 4,
  Maps = 5,
  MusicTracks = 6,
  Models = 7,
  Sprites = 8,
  Textures = 9,
  Binary = 10,
  MusicJingles = 11,
  ClientScripts = 12,
  Fonts = 13,
  MusicSamples = 14,
  MusicPatches = 15,
  Unused = 16,
  WorldMapGeography = 17,
  WorldMap = 18,
  WorldMapGrounds = 19,
  DbTableIndex = 21,
}

export enum ConfigType {
  Underlay = 1,
  IndentityKit = 3,
  Overlay = 4,
  Inventory = 5,
  Object = 6,
  Enum = 8,
  Npc = 9,
  Item = 10,
  Params = 11,
  Sequence = 12,
  SpotAnim = 13,
  VarBit = 14,
  VarClientString = 15,
  VarPlayer = 16,
  VarClient = 19,
  Hitsplat = 32,
  Healthbar = 33,
  Struct = 34,
  Area = 35,
  DbRow = 38,
  DbTable = 39,
}
