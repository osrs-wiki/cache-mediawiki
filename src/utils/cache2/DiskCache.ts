import {
  ArchiveData,
  CacheProvider,
  CacheVersion,
  FileProvider,
  hash,
  IndexData,
} from "./Cache";
import { Reader } from "./Reader";

export class DiskIndexData implements IndexData {
  public id!: number;
  public protocol!: number;
  public revision!: number;
  public compression!: number;
  public crc!: number;
  public named!: boolean;
  /** @internal */ archives: Map<number, ArchiveData> = new Map();
}

export class DiskCacheProvider implements CacheProvider {
  private data: Promise<Uint8Array | undefined>;
  private indexData: Map<number, Promise<DiskIndexData | undefined>> =
    new Map();
  private pointers: Map<number, Promise<Reader | undefined>> = new Map();

  public constructor(private readonly disk: FileProvider) {
    this.data = disk.getFile("main_file_cache.dat2");
    this.getPointers(255);
  }

  public async getIndex(index: number): Promise<DiskIndexData | undefined> {
    let id = this.indexData.get(index);
    if (!id) {
      this.indexData.set(
        index,
        (id = (async () => {
          const cdata = await this.getArchiveRaw(255, index);
          if (!cdata) {
            return undefined;
          }
          const ad = new ArchiveData(255, index);
          ad.addFile(0, 0);
          ad.compressedData = cdata;
          const r = new Reader(ad.getFile(0)!.data);

          const out = new DiskIndexData();
          out.id = index;
          const protocol = (out.protocol = r.u8());
          out.revision = protocol >= 6 ? r.i32() : -1;

          const flags = r.u8();
          const named = (out.named = !!(flags & 1));

          const numArchives = protocol <= 6 ? r.u16() : r.u32o16();
          const ams: ArchiveData[] = new Array(numArchives);
          for (let i = 0, id = 0; i < numArchives; i++) {
            id += protocol <= 6 ? r.u16() : r.u32o16();
            const v = new ArchiveData(index, id);
            ams[i] = v;
            out.archives.set(id, v);
          }

          if (named) {
            for (const am of ams) {
              am.namehash = r.i32();
            }
          }
          for (const am of ams) {
            am.crc = r.i32();
          }
          for (const am of ams) {
            am.revision = r.i32();
          }

          const numFileses = new Uint32Array(ams.length);
          for (let i = 0; i < numArchives; i++) {
            numFileses[i] = protocol <= 6 ? r.u16() : r.u32o16();
          }

          for (let i = 0; i < numArchives; i++) {
            const am = ams[i];
            const numFiles = numFileses[i];
            let id = 0;
            for (let i = 0; i < numFiles; i++) {
              id += protocol <= 6 ? r.u16() : r.u32o16();
              am.addFile(id, 0);
            }
          }

          if (named) {
            for (const am of ams) {
              for (const file of am.files.values()) {
                (<any>file).namehash = r.i32();
              }
            }
          }

          return out;
        })())
      );
    }
    return id;
  }

  private async getPointers(index: number): Promise<Reader | undefined> {
    let ptrs = this.pointers.get(index);
    if (!ptrs) {
      this.pointers.set(
        index,
        (ptrs = (async () => {
          const data = await this.disk.getFile("main_file_cache.idx" + index);
          return data && new Reader(data);
        })())
      );
    }
    return ptrs;
  }

  public async getArchives(index: number): Promise<number[] | undefined> {
    const idx = await this.getIndex(index);
    if (!idx) {
      return;
    }

    return Array.from(idx.archives.keys());
  }

  public async getArchive(
    index: number,
    archive: number
  ): Promise<ArchiveData | undefined> {
    const idx = await this.getIndex(index);
    if (!idx) {
      return;
    }
    const am = idx.archives.get(archive);
    if (!am) {
      return;
    }
    if (!am.compressedData) {
      const d = await this.getArchiveRaw(index, archive);
      if (!d) {
        return;
      }
      am.compressedData = d;
    }
    return am;
  }

  private async getArchiveRaw(
    index: number,
    archive: number
  ): Promise<Uint8Array | undefined> {
    const ptrs = await this.getPointers(index);
    if (!ptrs) {
      return undefined;
    }

    ptrs.offset = archive * 6;
    if (ptrs.remaining <= 0) {
      return undefined;
    }

    const len = ptrs.u24();
    let sector = ptrs.u24();

    const data = await this.data;
    const r = new Reader(data!);

    const headerSize = archive > 0xffff ? 10 : 8;
    const SECTOR_SIZE = 520;

    const fastRead = len + headerSize <= SECTOR_SIZE;

    const obuf = fastRead ? undefined : new Uint8Array(len);
    let readBytes = 0;
    let part = 0;
    for (; readBytes < len; ) {
      r.offset = sector * SECTOR_SIZE;
      const readArchive = headerSize == 10 ? r.i32() : r.u16();
      const readPart = r.u16();
      const nextSector = r.u24();
      const readIndex = r.u8();

      if (readArchive != archive) {
        throw new Error(
          `corrupted; read archive ${readArchive} wanted ${archive}`
        );
      }
      if (readPart != part) {
        throw new Error(`corrupted; read part ${readPart} wanted ${part}`);
      }
      if (readIndex != index) {
        throw new Error(`corrupted; read index ${readIndex} wanted ${index}`);
      }

      const sa = r.array(Math.min(SECTOR_SIZE - headerSize, len - readBytes));
      if (fastRead) {
        return sa;
      }
      obuf!.set(sa, readBytes);
      readBytes += sa.length;

      sector = nextSector;
      part++;
    }

    return obuf;
  }

  public async getArchiveByName(
    index: number,
    name: string | number
  ): Promise<ArchiveData | undefined> {
    const namehash = hash(name);

    const idx = await this.getIndex(index);
    if (!idx) {
      return;
    }

    for (const ar of idx.archives.values()) {
      if (ar.namehash === namehash) {
        if (!ar.compressedData) {
          const d = await this.getArchiveRaw(index, ar.archive);
          if (!d) {
            return;
          }
          ar.compressedData = d;
        }
        return ar;
      }
    }
  }

  public async getVersion(index: number): Promise<CacheVersion> {
    return {
      era: "osrs",
      indexRevision: (await this.getIndex(index))?.revision ?? 0,
    };
  }
}
