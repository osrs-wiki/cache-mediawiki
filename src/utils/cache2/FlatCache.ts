import {
  ArchiveData,
  CacheProvider,
  CacheVersion,
  FileProvider,
  hash,
  IndexData,
} from "./Cache";

export class FlatIndexData implements IndexData {
  public revision!: number;
  public compression!: number;
  public crc!: number;
  public named!: boolean;

  /**@internal*/ archives: Map<number, ArchiveData> = new Map();
  private constructor(public id: number) {}

  static async of(indexID: number, data: Uint8Array): Promise<FlatIndexData> {
    const LF = 0x0a;
    const EQ = 0x3d;
    const fid = new FlatIndexData(indexID);
    const td = new TextDecoder();
    let ar: ArchiveData | undefined;
    for (let i = 0; i < data.length; i++) {
      let eol = data.indexOf(LF, i);
      if (eol == -1) {
        eol = data.length;
      }
      const line = data.subarray(i, eol);
      i = eol;

      const kvs = line.indexOf(EQ);
      if (kvs == -1) {
        continue;
      }
      const key = td.decode(line.subarray(0, kvs));
      const value = td.decode(line.subarray(kvs + 1));

      if (key === "id") {
        fid.archives.set(~~value, (ar = new ArchiveData(indexID, ~~value)));
        continue;
      }

      if (!ar) {
        switch (key) {
          case "named":
            fid.named = value == "true";
            break;
          case "revision":
          case "compression":
          case "crc":
            fid[key] = ~~value;
            break;
        }
      } else {
        switch (key) {
          case "contents": {
            const res = await fetch(
              "data:application/octet-binary;base64," + value
            );
            const data = await res.arrayBuffer();
            ar!.compressedData = new Uint8Array(data);
            break;
          }
          case "file": {
            const [id, hash] = value.split("=");
            ar!.addFile(~~id, ~~hash);
            break;
          }
          case "namehash":
          case "revision":
          case "crc":
          case "compression":
            ar[key] = ~~value;
            break;
        }
      }
    }
    return fid;
  }

  getArchive(archive: number): ArchiveData | undefined {
    return this.archives.get(archive);
  }

  getArchiveByName(name: string | number): ArchiveData | undefined {
    const namehash = hash(name);
    for (const ar of this.archives.values()) {
      if (ar.namehash == namehash) {
        return ar;
      }
    }
  }
}

export class FlatCacheProvider implements CacheProvider {
  private indexes: Map<number, Promise<FlatIndexData | undefined>> = new Map();

  constructor(private disk: FileProvider) {}

  public async getIndex(index: number): Promise<FlatIndexData | undefined> {
    let idxp = this.indexes.get(index);
    if (!idxp) {
      this.indexes.set(
        index,
        (idxp = this.disk
          .getFile(index + ".flatcache")
          .then((data) => data && FlatIndexData.of(index, data)))
      );
    }
    return idxp;
  }

  public async getArchive(
    index: number,
    archive: number
  ): Promise<ArchiveData | undefined> {
    const idx = await this.getIndex(index);
    return idx?.getArchive(archive);
  }

  public async getArchives(index: number): Promise<number[] | undefined> {
    const idx = await this.getIndex(index);
    if (!idx) {
      return;
    }

    return Array.from(idx.archives.keys());
  }

  public async getArchiveByName(
    index: number,
    name: string | number
  ): Promise<ArchiveData | undefined> {
    const idx = await this.getIndex(index);
    return idx?.getArchiveByName(name);
  }

  public async getVersion(index: number): Promise<CacheVersion> {
    return {
      era: "osrs",
      indexRevision: (await this.getIndex(index))?.revision ?? 0,
    };
  }
}
