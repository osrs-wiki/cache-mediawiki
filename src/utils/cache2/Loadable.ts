import { CacheProvider } from "./Cache";
import { Reader } from "./Reader";

type LoadableType<I extends Loadable, ARGS extends unknown[]> = {
  decode(reader: Reader, ...args: ARGS): I;
  loadData(cache: CacheProvider, ...args: ARGS): Promise<Reader | undefined>;
};

type OrNumber<T extends unknown[]> = T;

export abstract class Loadable {
  public static load<I extends Loadable, ARGS extends unknown[]>(
    this: LoadableType<I, ARGS>,
    cache: CacheProvider | Promise<CacheProvider>,
    ...args: OrNumber<ARGS>
  ): Promise<I | undefined>;
  public static load<I extends Loadable, ARGS extends unknown[]>(
    this: LoadableType<I, ARGS>,
    reader: Reader | ArrayBufferView | ArrayBuffer,
    ...args: OrNumber<ARGS>
  ): I;
  public static load<I extends Loadable, ARGS extends unknown[]>(
    this: LoadableType<I, ARGS>,
    reader:
      | Reader
      | ArrayBufferView
      | ArrayBuffer
      | CacheProvider
      | Promise<CacheProvider>,
    ...args: OrNumber<ARGS>
  ): Promise<I | undefined> | I {
    if (reader instanceof ArrayBuffer || ArrayBuffer.isView(reader)) {
      reader = new Reader(reader);
    }
    if (reader instanceof Reader) {
      return this.decode(reader, ...args);
    } else {
      return (async () => {
        const data = await this.loadData(await reader, ...args);
        if (data) {
          return this.decode(data, ...args);
        }
      })();
    }
  }
}

export abstract class PerArchiveLoadable extends Loadable {
  public static async loadData(
    this: { index: number },
    cache: CacheProvider,
    id: number
  ): Promise<Reader | undefined> {
    const archive = await cache.getArchive(this.index, id as number);
    const version = await cache.getVersion(this.index);
    const data = archive?.getFile(0)?.data;
    return data ? new Reader(data, version) : undefined;
  }

  public static async all<I extends PerArchiveLoadable, ID extends number>(
    this: {
      index: number;
      decode(reader: Reader, id: ID): I;
    },
    cache0: CacheProvider | Promise<CacheProvider>
  ): Promise<I[]> {
    const cache = await cache0;
    const ids = await cache.getArchives(this.index);
    if (!ids) {
      return [];
    }

    const archives = await Promise.all(
      ids.map((id) => cache.getArchive(this.index, id))
    );
    const version = await cache.getVersion(this.index);

    return archives
      .filter((v) => v)
      .map((v) => {
        try {
          return this.decode(
            new Reader(v!.getFile(0)!.data, version),
            v!.archive as ID
          );
        } catch (e) {
          if (typeof e === "object" && e && "message" in e) {
            const ea = e as any;
            ea.message = v!.archive + ": " + ea.message;
          }
          throw e;
        }
      });
  }
}

export abstract class NamedPerArchiveLoadable extends PerArchiveLoadable {
  public static async loadByName<
    I extends PerArchiveLoadable,
    ID extends number
  >(
    this: {
      index: number;
      decode(reader: Reader, id: ID): I;
    },
    cache0: CacheProvider | Promise<CacheProvider>,
    name: string | number
  ): Promise<I | undefined> {
    const cache = await cache0;
    const ar = await cache.getArchiveByName(this.index, name);
    const version = await cache.getVersion(this.index);
    const data = ar?.getFile(0)?.data;
    if (data) {
      return this.decode(new Reader(data, version), ar!.archive as ID);
    }
  }
}

export class PerFileLoadable extends Loadable {
  public static async loadData(
    this: { index: number; archive: number },
    cache: CacheProvider,
    id: number
  ): Promise<Reader | undefined> {
    const archive = await cache.getArchive(this.index, this.archive);
    const version = await cache.getVersion(this.index);
    const data = archive?.getFile(id)?.data;
    return data ? new Reader(data, version) : undefined;
  }

  public static async all<I extends PerFileLoadable, ID extends number>(
    this: {
      index: number;
      archive: number;
      decode(reader: Reader, id: ID): I;
    },
    cache0: CacheProvider | Promise<CacheProvider>
  ): Promise<I[]> {
    const cache = await cache0;
    const ad = await cache.getArchive(this.index, this.archive);
    if (!ad) {
      return [];
    }

    const version = await cache.getVersion(this.index);

    return [...ad.getFiles().values()]
      .filter((v) => v.data.length > 1 && v.data[0] != 0)
      .map((v) => {
        try {
          return this.decode(new Reader(v.data, version), v.id as ID);
        } catch (e) {
          if (typeof e === "object" && e && "message" in e) {
            const ea = e as any;
            ea.message = v.id + ": " + ea.message;
          }
          throw e;
        }
      });
  }
}
