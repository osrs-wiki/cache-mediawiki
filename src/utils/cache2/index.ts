export * from "./LazyPromise";
export * from "./types";
