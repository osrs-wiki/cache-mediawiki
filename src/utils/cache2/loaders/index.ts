export * from "./Animation";
export * from "./Area";
export * from "./ClientScript";
export * from "./DBRow";
export * from "./Enum";
export * from "./GameVal";
export * from "./HealthBar";
export * from "./Hitsplat";
export * from "./Item";
export * from "./Location";
export * from "./LocationsDefinition";
export * from "./MapDefinition";
export * from "./NPC";
export * from "./Obj";
export * from "./Param";
export * from "./Position";
export * from "./Region";
export * from "./RegionMapper";
export * from "./SpotAnim";
export * from "./Sprite";
export * from "./Struct";
export * from "./Tile";
export * from "./Underlay";
export * from "./Varbit";
export * from "./VarPlayer";
export * from "./Widget";
