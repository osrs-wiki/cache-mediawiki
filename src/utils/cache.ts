import { DiskCacheProvider, FlatCacheProvider } from "@abextm/cache2";
import { readFile } from "fs/promises";

export type CacheSource = "github" | "local";
export type CacheFileType = "flat" | "disk";

export const getCacheProviderLocal = async (
  version: string,
  type: CacheFileType
) => {
  if (type === "disk") {
    return getCacheProviderDiskLocal(version);
  } else if (type === "flat") {
    return getCacheProviderFlatLocal(version);
  }
};

export const getCacheProviderDiskLocal = async (version: string) => {
  return new DiskCacheProvider({
    async getFile(name) {
      const ab = (await readFile(`./data/diskcache/${version}/${name}`)).buffer;
      return new Uint8Array(ab);
    },
  });
};

export const getCacheProviderFlatLocal = async (version: string) => {
  return new FlatCacheProvider({
    async getFile(name) {
      const ab = (await readFile(`./data/flatcache/${version}/${name}`)).buffer;
      return new Uint8Array(ab);
    },
  });
};

export const getCacheProviderGithub = async (version = "master") => {
  return new FlatCacheProvider({
    async getFile(name) {
      const req = await fetch(
        `https://raw.githubusercontent.com/abextm/osrs-cache/${version}/${name}`
      );
      if (!req.ok) {
        return;
      }
      const ab = await req.arrayBuffer();
      return new Uint8Array(ab);
    },
  });
};
