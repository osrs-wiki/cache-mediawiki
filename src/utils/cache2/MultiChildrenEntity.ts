import type { CacheProvider } from "./Cache";
import { PerFileLoadable } from "./Loadable";

/**
 * Base class for entities that have multiChildren functionality.
 * This abstract class extends PerFileLoadable and provides caching, deduplication,
 * and error handling for loading child entities.
 *
 * @example Usage with NPCs:
 * ```typescript
 * export class NPC extends MultiChildrenEntity<NPC, NPCID> {
 *   public multiChildren?: NPCID[] = [];
 *
 *   constructor(id: NPCID) {
 *     super(id);
 *   }
 *
 *   protected async loadChild(cache: Promise<CacheProvider>, childId: NPCID): Promise<NPC | null> {
 *     return NPC.load(cache, childId);
 *   }
 * }
 *
 * const npc = await NPC.load(cache, 14236);
 * const children = await npc.getMultiChildren(cache);
 * ```
 */
export abstract class MultiChildrenEntity<
  TEntity,
  TID extends number
> extends PerFileLoadable {
  /** Cache for loaded multiChildren entities */
  private _cachedMultiChildren?: TEntity[];

  constructor(public id: TID) {
    super();
  }

  /**
   * Abstract method that subclasses must implement to load a child entity
   * @param cache The cache provider
   * @param childId The ID of the child entity to load
   * @returns The loaded child entity or null if not found
   */
  protected abstract loadChild(
    cache: Promise<CacheProvider>,
    childId: TID
  ): Promise<TEntity | null>;

  /**
   * Get the multiChildren entities for this entity, loading them from cache if needed.
   * Results are cached to avoid repeated loading.
   * @param cache The cache provider to load children from
   * @returns Array of unique child entities (deduplicated by ID)
   */
  public async getMultiChildren(
    cache: Promise<CacheProvider>
  ): Promise<TEntity[]> {
    // Return cached result if available
    if (this._cachedMultiChildren !== undefined) {
      return this._cachedMultiChildren;
    }

    // If no multiChildren array, return empty array and cache it
    if (!this.multiChildren || this.multiChildren.length === 0) {
      this._cachedMultiChildren = [];
      return this._cachedMultiChildren;
    }

    try {
      // Deduplicate IDs before loading to avoid loading the same entity multiple times
      const uniqueIds = new Set<TID>();
      for (const childId of this.multiChildren) {
        if (childId > 0) {
          uniqueIds.add(childId);
        }
      }

      const childEntities: TEntity[] = [];

      // Load each unique child entity only once
      for (const childId of uniqueIds) {
        try {
          const childEntity = await this.loadChild(cache, childId);
          if (childEntity) {
            childEntities.push(childEntity);
          }
        } catch (e) {
          console.warn(
            `Failed to load child entity ${childId} for parent ${this.id}:`,
            e
          );
        }
      }

      // Cache and return the result (already deduplicated by loading unique IDs)
      this._cachedMultiChildren = childEntities;
      return this._cachedMultiChildren;
    } catch (error) {
      console.warn(
        `Failed to load multiChildren for entity ${this.id}:`,
        error
      );
      this._cachedMultiChildren = [];
      return this._cachedMultiChildren;
    }
  }

  /**
   * Check if this entity has multiChildren
   */
  public hasMultiChildren(): boolean {
    return Boolean(this.multiChildren && this.multiChildren.length > 0);
  }

  /**
   * Clear the multiChildren cache for this entity.
   * Useful for testing or when entity data changes.
   */
  public clearMultiChildrenCache(): void {
    this._cachedMultiChildren = undefined;
  }

  /**
   * Get the cached multiChildren without loading from cache.
   * Returns undefined if not yet cached.
   */
  public getCachedMultiChildren(): TEntity[] | undefined {
    return this._cachedMultiChildren;
  }

  // Abstract property that subclasses must define
  public abstract multiChildren?: TID[];
}
