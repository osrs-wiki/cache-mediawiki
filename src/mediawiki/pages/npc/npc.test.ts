import npcPageBuilder from "./npc";
import Context from "../../../context";

import { NPCID } from "@/utils/cache2";

describe("npcPageBuilder", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should build npc infobox without combat level", async () => {
    const builder = await npcPageBuilder({
      name: "name",
      combatLevel: 1,
      actions: ["action1", "action2"],
      id: 1 as NPCID,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any);
    expect(builder?.build()).toMatchSnapshot();
  });

  it("should build monster infobox with combat level", async () => {
    const builder = await npcPageBuilder({
      name: "name",
      combatLevel: 100,
      actions: ["action1", "action2"],
      id: 1 as NPCID,
      size: 2,
      hitpoints: 100,
      attack: 200,
      defence: 300,
      magic: 400,
      ranged: 500,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any);
    expect(builder?.build()).toMatchSnapshot();
  });

  it("should use beta id when Context.beta is true", async () => {
    const originalBeta = Context.beta;
    Context.beta = true;

    const builder = await npcPageBuilder({
      name: "name",
      combatLevel: 1,
      actions: ["action1", "action2"],
      id: 1 as NPCID,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any);
    expect(builder?.build()).toMatchSnapshot();

    Context.beta = originalBeta;
  });
});
