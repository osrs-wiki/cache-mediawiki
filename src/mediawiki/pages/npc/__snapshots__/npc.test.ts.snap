// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`npcPageBuilder should build monster infobox with combat level 1`] = `
"{{New Content}}
{{Infobox Monster
|name = name
|image = [[File:name.png|120px]]
|members = Yes
|combat = 100
|size = 2
|examine = 
|hitpoints = 100
|att = 200
|def = 300
|mage = 400
|range = 500
|id = 1
}}
[[File:name chathead.png|left]]
'''name'''"
`;

exports[`npcPageBuilder should build npc infobox without combat level 1`] = `
"{{New Content}}
{{Infobox Monster
|name = name
|image = [[File:name.png|120px]]
|members = Yes
|combat = 1
|examine = 
|id = 1
}}
[[File:name chathead.png|left]]
'''name'''"
`;

exports[`npcPageBuilder should build npc page with chathead when has chatheadModels 1`] = `
"{{New Content}}
{{Infobox Monster
|name = name
|image = [[File:name.png|120px]]
|members = Yes
|combat = 1
|examine = 
|id = 1
}}
[[File:name chathead.png|left]]
'''name'''"
`;

exports[`npcPageBuilder should build npc page with transcript 1`] = `
"{{New Content}}
{{Infobox Monster
|name = name
|image = [[File:name.png|120px]]
|members = Yes
|combat = 1
|examine = 
|id = 1
}}
[[File:name chathead.png|left]]
'''name'''

==Dialogue==
{{Hastranscript|npc}}
"
`;

exports[`npcPageBuilder should build npc page without chathead when no chatheadModels 1`] = `
"{{New Content}}
{{Infobox Monster
|name = name
|image = [[File:name.png|120px]]
|members = Yes
|combat = 1
|examine = 
|id = 1
}}
'''name'''"
`;

exports[`npcPageBuilder should strip HTML tags from NPC name in page content 1`] = `
"{{New Content}}
{{Infobox Monster
|name = Tornado
|image = [[File:Tornado.png|120px]]
|members = Yes
|combat = 1
|examine = 
|id = 1
}}
[[File:Tornado chathead.png|left]]
'''Tornado'''"
`;

exports[`npcPageBuilder should use beta id when Context.beta is true 1`] = `
"{{New Content}}
{{Infobox Monster
|name = name
|image = [[File:name.png|120px]]
|members = Yes
|combat = 1
|examine = 
|id = beta1
}}
[[File:name chathead.png|left]]
'''name'''"
`;
